spec.html

```diff
@@ -32497,173 +32497,107 @@ <h1>String.prototype.replace ( _searchValue_, _replaceValue_ )</h1>
-          1. Return the string-concatenation of _preserved_, _replacement_, and the substring of _string_ from _position_ + _searchLength_.
+          1. Return the string-concatenation of _preceding_, _replacement_, and _following_.
         </emu-alg>
         <emu-note>
           <p>The `replace` function is intentionally generic; it does not require that its *this* value be a String object. Therefore, it can be transferred to other kinds of objects for use as a method.</p>
         </emu-note>
 
-        <emu-clause id="sec-getsubstitution" type="abstract operation">
+        <emu-clause id="sec-getsubstitution" type="abstract operation" oldids="table-replacement-text-symbol-substitutions,table-45">
           <h1>
             GetSubstitution (
               _matched_: unknown,
               _str_: unknown,
               _position_: a non-negative integer,
               _captures_: unknown,
               _namedCaptures_: unknown,
-              _replacement_: unknown,
+              _replacer_: unknown,
             )
           </h1>
           <dl class="header">
           </dl>
+          <p>For the purposes of this abstract operation, a <em>decimal digit</em> is a code unit in the range 0x0030 (DIGIT ZERO) to 0x0039 (DIGIT NINE) inclusive.</p>
           <emu-alg>
             1. Assert: Type(_matched_) is String.
             1. Let _matchLength_ be the number of code units in _matched_.
             1. Assert: Type(_str_) is String.
             1. Let _stringLength_ be the number of code units in _str_.
             1. Assert: _position_ &le; _stringLength_.
-            1. Assert: _captures_ is a possibly empty List of Strings.
-            1. Assert: Type(_replacement_) is String.
+            1. Assert: _captures_ is a possibly empty List, each of whose elements is a String or *undefined*.
+            1. Assert: Type(_replacer_) is String.
+            1. Let _preceding_ be the substring of _str_ from 0 to _position_.
             1. Let _tailPos_ be _position_ + _matchLength_.
-            1. Let _m_ be the number of elements in _captures_.
-            1. Let _result_ be the String value derived from _replacement_ by copying code unit elements from _replacement_ to _result_ while performing replacements as specified in <emu-xref href="#table-replacement-text-symbol-substitutions"></emu-xref>. These `$` replacements are done left-to-right, and, once such a replacement is performed, the new replacement text is not subject to further replacements.
-            1. Return _result_.
+            1. If _tailPos_ &ge; _stringLength_, then
+              1. Let _following_ be the empty String.
+            1. Else,
+              1. Let _following_ be the substring of _str_ from _tailPos_.
+            1. Let _interpretNumericRef_ be a new Abstract Closure with parameters (_ref_) that captures _captures_ and performs the following steps when called:
+              1. Assert: _ref_ starts with *"$"*.
```

Comment:
There is no definition for "starts with"; would more precise language be better here (and in step 15)?
```suggestion
              1. Assert: The code unit at index 0 within _ref_ is 0x0024 (DOLLAR SIGN).
```

Reply:
> There is no definition for "starts with"

That's true, but it's also true of most other phrases that the spec uses to handle String values. (The only ones for which we define specific wording are "the string-concatenation of ..." and "the substring of ...", but phrases involving length and indexing could probably also be considered 'defined'.)

> would more precise language be better here (and in step 15)?
> `The code unit at index 0 within _ref_ is 0x0024 (DOLLAR SIGN)`

We could do that. (That would also take care of the kerning problem for `$'`.) The longer refs in step 15 could get pretty wordy, but rather than (e.g.)
```
If the code unit at index 0 within _replacerRemainder_ is 0x0024 (DOLLAR SIGN)
and the code unit at index 1 within _replacerRemainder_ is 0x0026 (AMPERSAND), then
```
we could say:
```
If the first two code units of _replacerRemainder_ are (respectively)
0x0024 (DOLLAR SIGN) and 0x0026 (AMPERSAND), then
```
(And again, there isn't a definition for "the first N code units of S", but it's used in `parseInt`, and we could define it if we felt it was necessary.)

We could also consider defining "starts with". Defining "A starts with B" where A and B are Strings would be fairly easy, but it's debatable whether that would handle step 15.e and 15.f:
```
15.e Else if _replacerRemainder_ starts with *"$"* followed by 2 (or more) decimal digits, then
15.f Else if _replacerRemainder_ starts with *"$"* followed by 1 decimal digit
```
because (e.g.) `*"$"* followed by 1 decimal digit` doesn't denote a String, it denotes a set of Strings. Or you can see it as an implicit existential quantification:
```
If there exists a String _x_ consisting of *"$"* followed by 1 decimal digit,
such that _replacerRemainder_ starts with _x_, then
```
I don't have a conclusion, it's just stuff that comes up when you pull on that thread.

Reply:
For my own part, I don't think we need to define "starts with": given that strings are defined to be sequences, I think readers should be able to understand what it means for one sequence to start with another.

Reply:
All of this is true, and there are many variations. I admit that I did have the `$'` problem on my mind as well, and that I don't have a better answer to 15.e than something like
```
            1. Repeat, while _replacerRemainder_ is not the empty String,
              1. Let _len_ be the length of _replacerRemainder_.
              1. If the code unit at index 0 within _replacerRemainder_ is not  0x0024 (DOLLAR SIGN), then
                1. Let _ref_ be the substring of _replacerRemainder_ from 0 to 1.
                1. Let _refReplacement_ be _ref_.
              …
              1. Else if _len_ &ge; 3 and the code units at index 1 and 2 within _replacerRemainder_ are both decimal digits, then
                1. Let _ref_ be the substring of _replacerRemainder_ from 0 to 3.
                1. Let _refReplacement_ be _interpretNumericRef_(_ref_).
```

Reply:
(I've continued to use "starts with".)